### PR TITLE
[Material] MaterialStateBuilder widget for custom interactive widgets.

### PR DESCRIPTION
**Pull request by johnsonmh**
## Description
                                                                           
`MaterialStateBuilder` is a convenience widget for widgets with custom interactive behavior.
The `MaterialStateBuilder` widget keeps track of the current `MaterialState`s using a combination of parameters and built in listener widgets. 

### Demo:
For example, take the following custom button in a flutter web project [(DartPad link)](https://dartpad.dev/embed-flutter.html?id=3d4c64a96dcc4e592f56b1bbd694a90b).

giphy

### Code:
Can be accomplished with the following sample app code [(DartPad link)](https://dartpad.dev/embed-flutter.html?id=3d4c64a96dcc4e592f56b1bbd694a90b).

```dart
import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';

void main() => runApp(_Home());

class _Home extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Center(
          child: MaterialStateBuilder(
            disabled: false,
            builder: (BuildContext context, Set<MaterialState> states) {
              final bool pressed = states.contains(MaterialState.pressed);
              final bool hovered = states.contains(MaterialState.hovered);

              return RaisedButton(
                onPressed: () {},
                color: Colors.blue,
                shape: RoundedRectangleBorder(
                  side: BorderSide(
                    color: pressed ? Colors.yellow : Colors.white,
                    width: hovered ? 2 : 1,
                  ),
                ),
                child: Text(
                  'Button',
                  style: TextStyle(
                    color: pressed ? Colors.yellow : Colors.white,
                    decoration: hovered ? TextDecoration.underline : null,
                    fontSize: 20.0,
                  ),
                ),
              );
            },
          ),
        ),
      ),
    );
  }
}
```

## Tests

I added the following tests:

* Tests the `MaterialStateBuilder` in all the states updates properly when enabling and disabling each state through parameters.
* Tests the `MaterialStateBuilder` uses listener widgets for hover/focus/pressed states.
* Tests that do combinations of states, and things like disabling a already hovered widget etc.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/material/chip.dart [hunk 1]
```diff
@@ -1711,112 +1686,118 @@ class _RawChipState extends State<RawChip> with TickerProviderStateMixin<RawChip
...
+            heightFactor: 1.0,
+          ),
+        );
+        return Semantics(
+          container: true,
+          selected: widget.selected,
+          enabled: canTap ? widget.isEnabled : null,
+          child: result,
+        );
+
```
Comment (rami-a) [inline packages/flutter/lib/src/material/chip.dart]: Extra line here

File: packages/flutter/lib/src/material/material_state.dart [hunk 2]
```diff
@@ -66,6 +68,163 @@ enum MaterialState {
 /// set of states.
 typedef MaterialPropertyResolver<T> = T Function(Set<MaterialState> states);
 
+/// A widget that uses [builder] to obtain its child widget.
```
Comment (rami-a) [inline packages/flutter/lib/src/material/material_state.dart]: I think the first sentence should be more specific and talk about the actual point of this widget.

File: packages/flutter/lib/src/material/material_state.dart [hunk 3]
```diff
@@ -66,6 +68,163 @@ enum MaterialState {
...
+      _updateState(MaterialState.pressed, widget.pressed);
+    _updateState(MaterialState.disabled, widget.disabled);
+    _updateState(MaterialState.selected, widget.selected);
+    _updateState(MaterialState.error, widget.error);
+  }
+
+  @override
+  void didUpdateWidget(MaterialStateBuilder oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (widget.hovered != null)
```
Comment (rami-a) [inline packages/flutter/lib/src/material/material_state.dart]: All of this function is the same as `initState`, perhaps consolidate it to a common private function they both call?

File: packages/flutter/lib/src/material/material_state.dart [hunk 4]
```diff
@@ -66,6 +68,163 @@ enum MaterialState {
...
+      result = MouseRegion(
+        onEnter: (_) => _updateState(MaterialState.hovered, true),
+        onExit: (_) => _updateState(MaterialState.hovered, false),
+        child: result,
+      );
+    }
+
+    if (widget.focused == null) {
+      result = Focus(
+        skipTraversal: true,
```
Comment (rami-a) [inline packages/flutter/lib/src/material/material_state.dart]: Why set this to true?

File: packages/flutter/lib/src/material/material_state.dart [hunk 5]
```diff
@@ -66,6 +68,163 @@ enum MaterialState {
...
+        onFocusChange: (bool value) => _updateState(MaterialState.focused, value),
+        child: result,
+      );
+    }
+
+    if (widget.pressed == null) {
+      result = GestureDetector(
+        excludeFromSemantics: true,
+        onForcePressStart: (_) => _updateState(MaterialState.pressed, true),
```
Comment (rami-a) [inline packages/flutter/lib/src/material/material_state.dart]: Are the force press callbacks necessary?

Also as an FYI, Apple is deprecating force/3D touch and the latest iphones don't even have the hardware for it.

File: packages/flutter/test/material/material_state_test.dart [hunk 6]
```diff
@@ -0,0 +1,378 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+@TestOn('!chrome') // whole file needs triage.
```
Comment (rami-a) [inline packages/flutter/test/material/material_state_test.dart]: What is this line about?

File: packages/flutter/test/material/material_state_test.dart [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+    await mouse.moveTo(tester.getCenter(find.byType(Text)));
+    await tester.pump(); // Register hover.
+    await tester.pump(); // Update Based on hover.
+    expect(find.text(defaultStateText), findsNothing);
+    expect(find.text(hoverText), findsOneWidget);
+
+    // Un-hover.
+    await mouse.moveTo(tester.getTopRight(find.byType(MaterialApp)));
+    await tester.pump(); // Register hover.
+    await tester.pump(); // Update Based on hover.
```
Comment (rami-a) [inline packages/flutter/test/material/material_state_test.dart]: So you have to do the double pump then? Did you find out why this is necessary?

File: packages/flutter/lib/src/material/chip.dart (hunk 1 above)
Comment (johnsonmh) [reply packages/flutter/lib/src/material/chip.dart]: Done.

File: packages/flutter/lib/src/material/material_state.dart (hunk 2 above)
Comment (johnsonmh) [reply packages/flutter/lib/src/material/material_state.dart]: Sure, how does this sound?
```
/// A widget that can be used to create widgets with custom interactive behavior,
/// such as a button that changes its label text style on hover/focus/press.
```

File: packages/flutter/lib/src/material/material_state.dart (hunk 3 above)
Comment (johnsonmh) [reply packages/flutter/lib/src/material/material_state.dart]: Done.

File: packages/flutter/lib/src/material/material_state.dart (hunk 4 above)
Comment (johnsonmh) [reply packages/flutter/lib/src/material/material_state.dart]: This is acting more as a "focus listener", we don't necessarily want to interrupt the focus node tree, but rather listen to children.

File: packages/flutter/lib/src/material/material_state.dart (hunk 5 above)
Comment (johnsonmh) [reply packages/flutter/lib/src/material/material_state.dart]: Ah didn't know this, removed.

File: packages/flutter/test/material/material_state_test.dart (hunk 6 above)
Comment (johnsonmh) [reply packages/flutter/test/material/material_state_test.dart]: I'm not totally sure, I copied it from another test file. I assume it's that tests for web need to be individually triaged before they are officially supported.

File: packages/flutter/test/material/material_state_test.dart (hunk 7 above)
Comment (johnsonmh) [reply packages/flutter/test/material/material_state_test.dart]: I believe hover and focus always require double pumps, the issue I was having before was the fact that buttons were now requiring an additional third pump for press, which I fixed by having the gesture detector be optional.

**Review by rami-a — APPROVED** [review]
LGTM

File: packages/flutter/test/material/material_state_test.dart (hunk 7 above)
Comment (johnsonmh) [reply packages/flutter/test/material/material_state_test.dart]: Yes as I understand it hover, focus, and press always require one frame to register and then one frame to update. My problem was that I was accidentally _adding_ an additional frame to buttons, I fixed that by making the `GestureDetector` optional in `MaterialStateBuilder`.

**Pull request closed without merging**